src/ast/ast.c

```diff
@@ -0,0 +1,328 @@
+    for(unsigned int i = 0; i < nerrors; i++) {
+        const cypher_parse_error_t *error = cypher_parse_result_get_error(result, i);
+
+        // Get the position of an error.
+        struct cypher_input_position errPos = cypher_parse_error_position(error);
+
+        // Get the error message of an error.
+        const char *errMsg = cypher_parse_error_message(error);
+
+        // Get the error context of an error.
+        // This returns a pointer to a null-terminated string, which contains a
+        // section of the input around where the error occurred, that is limited
+        // in length and suitable for presentation to a user.
+        const char *errCtx = cypher_parse_error_context(error);
+
+        // Get the offset into the context of an error.
+        // Identifies the point of the error within the context string, allowing
+        // this to be reported to the user, typically with an arrow pointing to the
+        // invalid character.
+        size_t errCtxOffset = cypher_parse_error_context_offset(error);
+
+        asprintf(&errorMsg, "errMsg: %s line: %u, column: %u, offset: %zu errCtx: %s errCtxOffset: %zu", errMsg, errPos.line, errPos.column, errPos.offset, errCtx, errCtxOffset);
+    }
+    return errorMsg;
+}
+
+// Recursively collect the names of all function calls beneath a node
+void AST_ReferredFunctions(const cypher_astnode_t *root, TrieMap *referred_funcs) {
+    cypher_astnode_type_t root_type = cypher_astnode_type(root);
+    if(root_type == CYPHER_AST_APPLY_OPERATOR) {
+        _consume_function_call_expression(root, referred_funcs);
+    } else {
+        unsigned int child_count = cypher_astnode_nchildren(root);
+        for(int i = 0; i < child_count; i++) {
+            const cypher_astnode_t *child = cypher_astnode_get_child(root, i);
+            AST_ReferredFunctions(child, referred_funcs);
+        }
+    }
+}
+
+
+// Retrieve the first instance of the specified clause in range, if any.
+const cypher_astnode_t* AST_GetClause(const AST *ast, cypher_astnode_type_t clause_type) {
+    uint clause_count = cypher_ast_query_nclauses(ast->root);
+    for (uint i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *child = cypher_ast_query_get_clause(ast->root, i);
+        if (cypher_astnode_type(child) == clause_type) return child;
+    }
+
+    return NULL;
+}
+
+/* Collect references to all clauses of the specified type in the query. Since clauses
+ * cannot be nested, we only need to check the immediate children of the query node. */
+uint AST_GetTopLevelClauses(const AST *ast, cypher_astnode_type_t clause_type, const cypher_astnode_t **matches) {
+    unsigned int num_found = 0;
+    uint clause_count = cypher_ast_query_nclauses(ast->root);
+    for (unsigned int i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *child = cypher_ast_query_get_clause(ast->root, i);
+        if (cypher_astnode_type(child) != clause_type) continue;
+
+        matches[num_found] = child;
```

Comment:
None; this call should be preceded by an `AST_GetClauseCount` call to size `matches` correctly.

I set it up this way so that `matches` could be a stack allocation instead of heap, but I don't know if the benefit of this is greater than the risk of misuse / ambiguity.

Reply:
I replaced this logic.